### PR TITLE
Added FFI bindings to lib.rs

### PR DESCRIPTION
**Pull request by ddefisher**

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -13,5 +13,6 @@ fn main() {
         .unwrap()
         .success());
     println!("cargo:rustc-flags=-L native={}", env::var("OUT_DIR").unwrap());
+    println!("cargo:rustc-flags=-l tinyfiledialogs -L target/debug/build/tinyfiledialogs-4949874712ebf6d8/out");
```
Comment (jdm) [inline build.rs]: I believe this should be equivalent to `println!("cargo:rustc-flags=-l tinyfiledialogs -L {}", env::var("OUT_DIR").unwrap())`, which is also future-proof.

File: src/lib.rs [hunk 2]
```diff
@@ -1,6 +1,52 @@
+extern crate libc;
+use libc::c_char;
+use libc::c_uchar;
+use libc::c_int;
+
 #[cfg(test)]
 mod test {
```
Comment (jdm) [inline src/lib.rs]: We can remove this block now.

File: src/lib.rs [hunk 3]
```diff
@@ -1,6 +1,52 @@
...
+    pub fn tinyfd_inputBox (
+        aTitle: *const c_char ,
+        aMessage: *const c_char,
+        aDefaultInput: *const c_char) -> *const c_char;
+
+    pub fn tinyfd_saveFileDialog (
+        aTitle: *const c_char,
+        aDefaultPathAndFile: *const c_char,
+        aNumOfFilterPatterns: c_int,
+        //aFilterPatterns: *mut *const c_char,
```
Comment (jdm) [inline src/lib.rs]: Let's remove this.

File: src/lib.rs [hunk 4]
```diff
@@ -1,6 +1,52 @@
...
+        aFilterPatterns: *const c_char,
+        aSingleFilterDescription: *const c_char) -> *const c_char;
+
+    pub fn tinyfd_openFileDialog (
+        aTitle: *const c_char,
+        aDefaultPathAndFile: *const c_char,
+        aNumOfFilterPatterns: c_int,
+        //aFilterPatterns: *mut *const c_char,
```
Comment (jdm) [inline src/lib.rs]: This too.

File: src/lib.rs [hunk 5]
```diff
@@ -1,6 +1,52 @@
...
+
+    pub fn tinyfd_selectFolderDialog (
+        aTitle: *const c_char,
+        aDefaultPath: *const c_char) -> *const c_char;
+
+    pub fn tinyfd_colorChooser (
+        aTitle: *const c_char,
+        aDefaultHexRGB: *const c_char,
+        aDefaultRGB: [c_uchar ; 3],
+        aoResultRGB: [c_uchar; 3]) -> *const c_char;
```
Comment (jdm) [inline src/lib.rs]: I'm pretty sure this should be `*const c_char` as well (or possibly `&[c_uchar; 3]` would also work and be safer), since `T[N]` as a function argument in C is actually a pointer.

**Comment by jdm** [comment]
Apart from those notes, this looks great! We should include the Rust example in an `examples` directory, as well.

**Comment by ddefisher** [comment]
I have added the rust example I am not sure if that is the best way to call the C functions.

File: examples/tinytest/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
+extern crate tinyfiledialogs;
+extern crate libc;
+
+use std::ffi::{CString, CStr};
+use libc::c_int;
+use libc::c_uchar;
+
+
+fn message_box(title: &CStr, message: &CStr, dialog_type: &CStr, icon: &CStr, button: c_int) {
+
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: nit: remove the extra newline after each `fn` line.

File: examples/tinytest/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    unsafe{
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: nit: add a space before each {

File: examples/tinytest/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: nit: remove the extra newline after each `unsafe` line.

File: examples/tinytest/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        tinyfiledialogs::tinyfd_messageBox(
+            title.as_ptr(),
+            message.as_ptr(),
+            dialog_type.as_ptr(),
+            icon.as_ptr(),
+            button);
+    }
+}
+fn input_box(title: &CStr, message: &CStr, default: &CStr) {
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: nit: add a newline before each `fn` line.

File: examples/tinytest/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+    unsafe{
+
+        tinyfiledialogs::tinyfd_inputBox(title.as_ptr(),message.as_ptr(),default.as_ptr());
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: nit: add a space after each , when there are multiple arguments on the same line.

File: src/lib.rs [hunk 11]
```diff
@@ -41,6 +39,6 @@ extern {
     pub fn tinyfd_colorChooser (
         aTitle: *const c_char,
         aDefaultHexRGB: *const c_char,
-        aDefaultRGB: [c_uchar ; 3],
-        aoResultRGB: [c_uchar; 3]) -> *const c_char;
+        aDefaultRGB: &[c_uchar ; 3],
+        aoResultRGB: &[c_uchar; 3]) -> *const c_char;
```
Comment (jdm) [inline src/lib.rs]: Turns out I was wrong and this isn't safe either (since in Rust `&[T; N]` is represented as `(*const T, usize)`). This will need to be `*const c_uchar` instead.

**Comment by jdm** [comment]
Looks good apart from the mostly stylistic comments! I've also made a pull request to your branch to turn on automated tests using TravisCI, so please merge that :)

File: examples/tinytest/src/main.rs [hunk 12]
```diff
@@ -124,14 +113,12 @@ fn main()
 
     let color_title                      = CString::new("Color Choose").unwrap();
     let color_default_hex                = CString::new("Color Choose").unwrap();
-    let color_default_RGB : [c_uchar ; 3] = [250,0,0];
-    let color_result_RGB  : [c_uchar ; 3] = [100,0,0];
+    let color_default_rgb                = CString::new("FFF").unwrap();
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: This doesn't seem to match the documentation at https://github.com/jdm/tinyfiledialogs/blob/master/libtinyfiledialogs/tinyfiledialogs.h#L128 . Why this change?

File: examples/tinytest/src/main.rs [hunk 13]
```diff
@@ -124,14 +113,12 @@ fn main()
 
     let color_title                      = CString::new("Color Choose").unwrap();
     let color_default_hex                = CString::new("Color Choose").unwrap();
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: The documentation suggests this should be a valid hex string like "#FF0000".

File: src/lib.rs [hunk 14]
```diff
@@ -1,6 +1,43 @@
...
+        aDefaultPathAndFile: *const c_char,
+        aNumOfFilterPatterns: c_int,
+        aFilterPatterns: *const c_char,
+        aSingleFilterDescription: *const c_char) -> *const c_char;
+
+    pub fn tinyfd_openFileDialog (
+        aTitle: *const c_char,
+        aDefaultPathAndFile: *const c_char,
+        aNumOfFilterPatterns: c_int,
+        aFilterPatterns: *const c_char,
```
Comment (jdm) [inline src/lib.rs]: `*const *const c_char`

File: src/lib.rs [hunk 15]
```diff
@@ -39,6 +38,6 @@ extern {
     pub fn tinyfd_colorChooser (
         aTitle: *const c_char,
         aDefaultHexRGB: *const c_char,
-        aDefaultRGB: &[c_uchar ; 3],
-        aoResultRGB: &[c_uchar; 3]) -> *const c_char;
+        aDefaultRGB: *const c_char,
+        aoResultRGB: *const c_char) -> *const c_char;
```
Comment (jdm) [inline src/lib.rs]: `*const c_uchar`

File: src/lib.rs [hunk 16]
```diff
@@ -1,6 +1,43 @@
...
+    pub fn tinyfd_inputBox (
+        aTitle: *const c_char ,
+        aMessage: *const c_char,
+        aDefaultInput: *const c_char) -> *const c_char;
+
+    pub fn tinyfd_saveFileDialog (
+        aTitle: *const c_char,
+        aDefaultPathAndFile: *const c_char,
+        aNumOfFilterPatterns: c_int,
+        aFilterPatterns: *const c_char,
```
Comment (jdm) [inline src/lib.rs]: `*const *const c_char`

File: examples/tinytest/src/main.rs (hunk 12 above)
Comment (jdm) [inline examples/tinytest/src/main.rs]: We should really be passing `&[100, 0, 0]` and using `color_default_rgb.as_ptr()` when calling the FFI function.

**Comment by jdm** [comment]
Looking good apart from the most recent comments!

File: examples/tinytest/src/main.rs [hunk 17]
```diff
@@ -2,66 +2,89 @@ extern crate tinyfiledialogs;
...
+                    filter_pattern_one: &CStr, des: &CStr) -> *const c_char {
+    let filter_patterns : [&CStr; 1] = [filter_pattern_one];
     unsafe {
-        tinyfiledialogs::tinyfd_saveFileDialog(
+        let result: *const c_char = tinyfiledialogs::tinyfd_saveFileDialog(
             title.as_ptr(),
             path.as_ptr(),
             num_patterns,
-            filter_patterns.as_ptr(),
+            filter_patterns.iter().map(|c| c.as_ptr()).collect::<Vec<*const c_char>>().as_ptr(),
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: This needs to be in a local variable in order to be safe; otherwise the pointer is pointing at memory that is freed as soon as the collect() expression goes out of scope.

File: examples/tinytest/src/main.rs [hunk 18]
```diff
@@ -2,66 +2,89 @@ extern crate tinyfiledialogs;
...
+                    des: &CStr, multi_select: c_int) -> *const c_char {
+    let filter_patterns : [&CStr; 1] = [filter_pattern_one];
     unsafe {
-        tinyfiledialogs::tinyfd_openFileDialog(
+        let result: *const c_char = tinyfiledialogs::tinyfd_openFileDialog(
             title.as_ptr(),
             path.as_ptr(),
             num_patterns,
-            filter_patterns.as_ptr(),
+            filter_patterns.iter().map(|c| c.as_ptr()).collect::<Vec<*const c_char>>().as_ptr(),
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: Same as previous.

**Comment by jdm** [comment]
Fix those comments and I'll merge this :)

**Comment by ddefisher** [comment]
jdm I have moved the map as_ptr into a local variable

File: examples/tinytest/src/main.rs [hunk 19]
```diff
@@ -33,12 +33,13 @@ fn save_file_dialog(title: &CStr, path: &CStr,
                     num_patterns: c_int,
                     filter_pattern_one: &CStr, des: &CStr) -> *const c_char {
     let filter_patterns : [&CStr; 1] = [filter_pattern_one];
+    let ptr_filter_patterns = filter_patterns.iter().map(|c| c.as_ptr()).collect::<Vec<*const c_char>>().as_ptr();
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: Sorry, this is still unsafe. The as_ptr() is what makes the Vec go out of scope; it should be moved to the actual argument to the function call instead.

File: examples/tinytest/src/main.rs [hunk 20]
```diff
@@ -49,12 +50,13 @@ fn open_file_dialog(title: &CStr, path: &CStr,
                     num_patterns: c_int, filter_pattern_one: &CStr,
                     des: &CStr, multi_select: c_int) -> *const c_char {
     let filter_patterns : [&CStr; 1] = [filter_pattern_one];
+    let ptr_filter_patterns = filter_patterns.iter().map(|c| c.as_ptr()).collect::<Vec<*const c_char>>().as_ptr();
```
Comment (jdm) [inline examples/tinytest/src/main.rs]: Same.

**Comment by ddefisher** [comment]
moved vector as_ptr method to function call argument

**Comment by jdm** [comment]
Thanks!
